app/services/submissions.server.ts

```diff
@@ -1,76 +1,141 @@
-  const { serviceRequestId, laborMarketAddress, ...data } = submission;
-  const newSubmission = await prisma.submission.upsert({
-    where: { contractId_laborMarketAddress: { contractId: submission.contractId, laborMarketAddress } },
-    update: data,
-    create: {
-      serviceRequestId,
-      laborMarketAddress,
-      ...data,
-    },
-  });
-  return newSubmission;
+const searchParams = (params: SubmissionSearch): Parameters<typeof mongo.submissions.find>[0] => {
+  return {
+    valid: true,
+    ...(params.laborMarketAddress ? { address: params.laborMarketAddress } : {}),
+    ...(params.serviceRequestId ? { address: params.serviceRequestId } : {}),
+    ...(params.q ? { $text: { $search: params.q, $language: "english" } } : {}),
+  };
 };
 
+/**
+ * Finds a Submission by its ID.
+ * @param {String} id - The ID of the submission.
+ * @returns - The Submission or null if not found.
+ */
+export const findSubmission = async (id: string, laborMarketAddress: string) => {
+  return mongo.submissions.findOne({ id, address: laborMarketAddress, valid: true });
+};
+
+/**
+ * Counts the number of Submissions on a particular service request.
+ * @param {SubmissionSearch} params - The search parameters.
+ * @returns {number} - The number of submissions that match the search.
+ */
+export const countSubmissionsOnServiceRequest = async (serviceRequestId: string) => {
+  return mongo.submissions.countDocuments({ serviceRequestId, valid: true });
+};
+
+/**
+ * Create a new SubmissionDoc from a TracerEvent.
+ */
+export const indexSubmission = async (event: TracerEvent) => {
+  const contract = LaborMarket__factory.connect(event.contract.address, nodeProvider);
+  const requestId = z.string().parse(event.decoded.inputs.requestId);
+  const submission = await contract.serviceSubmissions(requestId, { blockTag: event.block.number });
+  const appData = await fetchIpfsJson(submission.uri)
+    .then(submissionMetaDataSchema.parse)
+    .catch(() => null);
+
+  const currentSubmissionCount = await countSubmissionsOnServiceRequest(requestId);
+  console.log("CURRENT SUBMISSION COUNT", currentSubmissionCount);
+
+  console.log("REQUEST ID", submission.requestId.toString());
+  const isValid = appData !== null;
+  // Build the document, omitting the serviceRequestCount field which is set in the upsert below.
+  const doc: Omit<SubmissionDoc, "reviewCount"> = {
+    id: (currentSubmissionCount + 1).toString(),
+    laborMarketAddress: event.contract.address,
+    serviceRequestId: requestId,
+    valid: isValid,
+    reviewed: submission.reviewed,
+    submissionUrl: appData?.submissionUrl ? appData.submissionUrl : null,
+    indexedAt: new Date(),
+    configuration: {
+      requester: submission.serviceProvider,
+      uri: submission.uri,
+    },
+    appData,
+  };
+
+  if (isValid) {
+    await mongo.serviceRequests.updateOne(
+      { id: doc.serviceRequestId },
+      {
+        $inc: {
+          submissionCount: 1,
```

Comment:
Thinking we should rename this to make it more obvious
serviceRequests.submissionCount  -> serviceRequests.validSubmissionCount